### DISCUSSION TITLE
How to avoid Swift concurrency warnings with GRDB 6?

### DISCUSSION THREAD

**Discussion by groue** [groue/GRDB.swift#1509] [Q&A] [answered]
The Swift 5.10 compiler and concurrency checks can trigger warnings in your GRDB-powered application.

The future GRDB 7 aims at solving those warnings. Meanwhile, you might wonder how to fix compiler diagnostics with GRDB 6.

**Please ask your questions below, so that we can share answers and best practices.**

**Comment by groue** [comment]
## Warnings in associations declarations

The Swift compiler can warn when you declare a association between record types:

```swift
import GRDB

extension Author: Codable, FetchableRecord, PersistableRecord {
  // ⚠️ Static property 'books' is not concurrency-safe because it is not
  // either conforming to 'Sendable' or isolated to a global actor;
  // this is an error in Swift 6
  static let books = hasMany(Book.self)
}
```

The reason for this warning is that GRBD associations do not conform to the `Sendable` protocol.

Associations will be sendable in the next major version, GRDB 7. Until, then please apply one of those techniques:

1. (SOON) Import GRDB with `@preconcurrency`:

    ```swift
    @preconcurrency import GRDB
    
    extension Author: Codable, FetchableRecord, PersistableRecord {
      static let books = hasMany(Book.self)
    }
    ```

    😬 Unfortunately, there is a [bug](https://github.com/apple/swift/issues/72187) in the Swift 5.10 compiler (Xcode 15.3) that prevents this technique from removing the warning 🤦

2. Declare the association as `nonisolated(unsafe)`

    ```swift
    import GRDB
    
    extension Author: Codable, FetchableRecord, PersistableRecord {
      // Declare associations as nonisolated until GRDB7.
      // This is safe according to <https://github.com/groue/GRDB.swift/discussions/1509>
      nonisolated(unsafe) static let books = hasMany(Book.self)
    }
    ```

    The "unsafe" word sounds scary, but the `belongsTo(Author.self)` association is completely safe. You will not have any problem at runtime. The type of this association is not marked as `Sendable`, which explains the compiler warning. But this particular instance of association has all the runtime properties of a sendable value. It won't blow up your program, 100% guaranteed. But you have to trust my words.

    <details><summary>How to be really unsafe</summary>

    It is possible to create an unsafe association, but it is really involved and nobody does that. For example, the association below is unsafe because it embeds a non-sendable type, and will invoke it from the various threads that connect to the database:
    
    ```swift
    extension Author {
      // UNSAFE ASSOCIATION
      static func wackyBooks(_ value: NonSendableType) -> HasManyAssociation<Author, Book> {
        Author.hasMany(Book.self).filterWhenConnected { db in
          Column("name") == value.description
        }
      }
    }
    ```
    
    Since you do not write code like that, your associations can safely be declared as `nonisolated(unsafe`.
    
    </details>

**Comment by groue** [comment, marked as answer]
GRDB 7 solves the warnings of GRDB 6.

**Comment by Mehul721** [comment]
<img width="1504" alt="Screenshot 2024-10-06 at 4 48 29 PM" src="https://github.com/user-attachments/assets/6394f7f5-f81e-42eb-9b9c-604e07a570c2">
Hi.I have updated the grdb to 7.I am still getting these errors which suddenly disappear as soon as I open that file to look into those errors.But they comeback again as soon as I re-build the proejct.Can you help me here?Thank you.

**Reply by groue** [reply]
Hello @Mehul721. The code you show does not look related with GRDB. Please refer to general Swift 6 documentation, such as the [Migrating to Swift 6](https://www.swift.org/migration/documentation/migrationguide) Apple guide.
